### DISCUSSION TITLE
How to add custom FHIR definitions?

### DISCUSSION THREAD

**Discussion by marcoadasilvaa** [medplum/medplum#5383] [Q&A]
Hi, thanks for contributing to the community with this excellent tool.

As context, I have been reviewing the code to add a regional version of FHIR.

What would be the correct approach to include new: CapabilityStatement, CodeSystem, StructureDefinition, and ValueSet? I understand that everything starts in the package/definition and then continues with the package/generator, which should generate the corresponding fhirtypes and migrations. However, I have made several attempts without success so far.

**Comment by codyebberson** [comment]
Hi @marcoadasilvaa - these are good questions.

Adding `CodeSystem` and `ValueSet` resources should "just work".  You can create the FHIR resources using the normal FHIR REST API.  If the `CodeSystem` has many codes (i.e., 1000+), then you can use our custom `CodeSystem/$import` operation.  There is some rough documentation here: https://www.medplum.com/docs/contributing/terminology-architecture#codesystemimport

Adding custom `StructureDefinition` resources is more complicated.  You can create the FHIR resource using the normal FHIR REST API, especially for FHIR Profile resources.  For creating "base" types, that is not well supported.

We currently do not support overriding the `/$metadata` `CapabilityStatement`.  At present, the `CapabilityStatement` is generated dynamically in [`metadata.ts`](https://github.com/medplum/medplum/blob/main/packages/server/src/fhir/metadata.ts).

Your questions are reasonable, and this work is on the roadmap, it's just not fully developed today.  If you'd like to discuss more, please feel free to reach out to <hello@medplum.com> or schedule some time at <https://cal.com/medplum/30min>
